topics/data-science/tutorials/snakemake/tutorial.md

```diff
@@ -0,0 +1,1714 @@
+> >
+> >  rule fastqc:
+> >  	input:
+> > -		"{sample}.fq"
+> > +		"reads/{sample}.fq.gz"
+> >  	output:
+> > -		"{sample}_fastqc.html"
+> > +		"fastqc/{sample}_fastqc.html"
+> >  	conda:
+> >  		"envs/fastqc.yaml"
+> >  	log:
+> > @@ -31,13 +31,13 @@ rule fastqc:
+> >
+> >  rule trimmomatic:
+> >  	input:
+> > -		r1="{sample}_1.fq",
+> > -		r2="{sample}_2.fq"
+> > +		r1="reads/{sample}_1.fq",
+> > +		r2="reads/{sample}_2.fq"
+> >  	output:
+> > -		o1="{sample}_1.trim.fq",
+> > -		o2="{sample}_2.trim.fq",
+> > -		o1un="{sample}_1un.trim.fq",
+> > -		o2un="{sample}_2un.trim.fq"
+> > +		o1="trim/{sample}_1.trim.fq",
+> > +		o2="trim/{sample}_2.trim.fq",
+> > +		o1un="trim/{sample}_1un.trim.fq",
+> > +		o2un="trim/{sample}_2un.trim.fq"
+> >  	conda:
+> >  		"envs/trimmomatic.yaml"
+> >  	log:
+> > @@ -53,9 +53,9 @@ rule trimmomatic:
+> >
+> >  rule indexgenome:
+> >  	input:
+> > -		"GCA_000017985.1_ASM1798v1_genomic.fna"
+> > +		"reference/GCA_000017985.1_ASM1798v1_genomic.fa"
+> >  	output:
+> > -		"GCA_000017985.1_ASM1798v1_genomic.fna.bwt"
+> > +		"reference/GCA_000017985.1_ASM1798v1_genomic.fa.bwt"
+> >  	conda:
+> >  		"envs/bwa.yaml"
+> >  	log:
+> > @@ -66,11 +66,11 @@ rule indexgenome:
+> >
+> >  rule align:
+> >  	input:
+> > -		r1="{sample}_1.trim.fq",
+> > -		r2="{sample}_1.trim.fq",
+> > -		index="{genome}.fna.bwt"
+> > +		r1="trim/{sample}_1.trim.fq",
+> > +		r2="trim/{sample}_1.trim.fq",
+> > +		index="reference/{genome}.fna.bwt"
+> >  	output:
+> > -		"{genome}/{sample}.bam"
+> > +		"alignments/{genome}/{sample}.bam"
+> >  	conda:
+> >  		"envs/bwa.yaml"
+> >  	log:
+> > @@ -78,5 +78,5 @@ rule align:
+> >  		out="logs/samtools.{genome}.{sample}.out",
+> >  		err="logs/samtools.{genome}.{sample}.err"
+> >  	shell:
+> > -		"bwa mem {wildcard.genome}.fna {input.r1} {input.r2} 2>{log.bwaerr} | "
+> > +		"bwa mem reference/{wildcards.genome}.fna.gz {input.r1} {input.r2} 2>{log.bwaerr} | "
+> >  		"samtools sort -O bam -o {output} >{log.out} 2>{log.err}"
+> > ```
+> {: .solution}
+{: .hands_on}
+
+Notice how while updating data location, we only had to update the input and output boxes of the Snakefile, we didn't need to change any commands because there we used `{input}` or `{output}` which are automatically templated for us. Except for the last command where we had to use a wildcard and manually construct the path to work around the issue of 'fake' inputs where the tool expected a filename that was the common subset of all of the files it would produce.
+
+
+### Set a Default Task
+
+In the Makefile we had an `all` rule which was the first and default action to take. Let's reproduce that.
+
+> ### {% icon hands_on %} Hands-on: Add an all task
+> It should build `alignments/GCA_000017985.1_ASM1798v1_genomic/SRR2584863.bam`
+>
+> > ### {% icon solution %} Solution
+> > We add this to the very top of our file.
+> > ```diff
+> > --- a/Snakefile
+> > +++ b/Snakefile
+> > @@ -1,3 +1,7 @@
+> > +rule all:
+> > +	input:
```

Comment:
i do understand the concept of rule all:
but why is ```input: ``` used and not ```outpu:``` syntax wise?

Reply:
because rule all is taking in some other outputs as an input to it. If we used `outputs`, rule all would do nothing (it has no pre-requisite steps), and since it produces nothing by itself, snakemake would see that the outputs you expected have not been created and flag it as an error.